types/share_splitting.go

```diff
@@ -2,11 +2,62 @@ package types
 
 import (
 	"bytes"
+	"fmt"
+	"sort"
 
 	"github.com/celestiaorg/nmt/namespace"
 	"github.com/tendermint/tendermint/pkg/consts"
 )
 
+// MessageShareWriter lazily merges messages into shares that will eventually be
+// included in a data square. It also has methods to help progressively count
+// how many shares the messages written take up.
+type MessageShareWriter struct {
```

Comment:
Why do we need this ShareWriter if we have a contiguous one?

Reply:
Also, don't see it being used somewhere

Reply:
Message are encoded contiguously, they are padded so that any share will only include one message.

Its not being used here, although I think it could be. It is used in the app tho

https://github.com/celestiaorg/lazyledger-app/blob/e18d8d2301a96702e1bf684735a3620eb059b12f/app/square_writer.go#L106-L121

Reply:
I'll look into if we should use here more and get back to this thread

Reply:
ahh, so the current code for splitting messages into shares is so simple that I didn't think it would improve it to use it here

https://github.com/celestiaorg/celestia-core/blob/7bb534fd06764cf2e508c2875bf526574a733cd4/types/block.go#L1210-L1221

Reply:
we can tho if we think that is worth it